temporalio/converter.py

```diff
@@ -0,0 +1,362 @@
+from typing import Any, Optional, Tuple, Type
+
+import google.protobuf.json_format
+import google.protobuf.message
+import google.protobuf.symbol_database
+
+import temporalio.api.common.v1
+
+
+class DataConverter(ABC):
+    """Base converter to/from multiple payloads/values."""
+
+    @abstractmethod
+    async def encode(self, values: list[Any]) -> list[temporalio.api.common.v1.Payload]:
+        """Encode values into payloads.
+
+        Args:
+            values: Values to be converted.
+
+        Returns:
+            Converted payloads. Note, this does not have to be the same number
+            as values given, but at least one must be present.
+
+        Raises:
+            Exception: Any issue during conversion.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def decode(
+        self,
+        payloads: list[temporalio.api.common.v1.Payload],
+        type_hints: Optional[list[Type]],
+    ) -> list[Any]:
+        """Decode payloads into values.
+
+        Args:
+            payloads: Payloads to convert to Python values.
+            type_hints: Types that are expected if any. This may not have any
+                types if there are no annotations on the target. If this is
+                present, it must have the exact same length as payloads even if
+                the values are just "object".
+
+        Return:
+            Collection of Python values. Note, this does not have to be the same
+            number as values given, but at least one must be present.
+
+        Raises:
+            Exception: Any issue during conversion.
+        """
+        raise NotImplementedError
+
+
+class PayloadConverter(ABC):
+    """Base converter to/from single payload/value."""
+
+    @abstractmethod
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """Encode a single value to a payload or None.
+
+        Args:
+            value: Value to be converted.
+
+        Returns:
+            Payload of the value or None if unable to convert.
+        """
+        return None
+
+    @abstractmethod
+    async def decode(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Tuple[Any, bool]:
+        """Decode a single payload to a Python value if able.
+
+        Args:
+            payload: Payload to convert to Python value.
+            type_hints: Type that is expected if any. This may not have a type
+                if there are no annotations on the target.
+
+        Return:
+            A tuple with the first value as the Python value and the second as
+            whether it could be converted or not. If the payload can be
+            converted, it is the first value of the tuple and the second value
+            is True. If the payload cannot be converted, the first value is
+            undefined and the second value is False.
+        """
+        return (None, False)
```

Comment:
Maybe it makes sense to expect users to throw if a payload cannot be converted.
I feel like this is a more pythonic approach.

Reply:
Also, this shouldn't typically happen since the composite converter looks at the payload header before dispatching.

Reply:
> I feel like this is a more pythonic approach.

I don't. From what I read, they much prefer returns than exceptions for control flow. Exceptions are more expensive.

> Also, this shouldn't typically happen since the composite converter looks at the payload header before dispatching.

This version does not do that so you could have a converter handle multiple encodings, but other SDKs do. Shall I make it do that?

Reply:
I'm changing to exception based

Reply:
Changed